stackrox/stackrox
# ROX-11605: Display clusterIdToRetentionInfo in clusters list and cluster panel

## Description

Follow up #2182

1. Adapt `ClusterDeletion` component from `CertificateExpiration` but without icon
    * **Not applicable** if `clusterRetentionInfo === null`
    * **in D days** if `'daysUntilDeletion' in clusterRetentionInfo`
    * **Excluded from deletion** if `isExcluded` 
    * **Deletion is turned off** if `!isExcluded`
2. Clusters table
    * Render **Cluster Deletion** column if feature flag is enabled
    * Omit row actions unless role has READ_WRITE_ACCESS for Cluster resource
3. Cluster side panel
    * Render **Cluster Deletion** column if feature flag is enabled

### Changed files

1. Edit cypress/fixtures/clusters/health.json
    * Add `clusterIdToRetentionInfo` property

2. Edit cypress/helpers/clusters.js
    * Factor out `visitClustersWithFixture` helper function
    * Remove clusters request from route matcher map for `visitClusterById` function for forward compatibility with cluster page
    * Add `clusterRetentionInfo` prop in `visitClusterByNameWithFixture` and `visitClusterByNameWithFixtureMetadataDatetime` functions

3. Add cypress/integration/clusters subfolder and move clusters.test.js into it

4. Add cypress/integration/clusters/clusterDeletion.test.js

5. Edit src/Containers/Clusters/ClusterEditForm.tsx
    * Add `clusterRetentionInfo` prop

6. Edit src/Containers/Clusters/ClustersSidePanel.tsx
    * Import `Cluster` from frontend proto type and add `as unknown as Cluster` cast to get rid of `Partial<Cluster> | null` type
    * Replace `getClusterById` with `fetchClusterWithRetentionInformation` function call
    * Destructure `clusterRetentionInfo` property from cluster response
    * Add comment about future code to prevent possible edge case from not updating retention info after save

7. Edit src/Containers/Clusters/ClustersTablePanel.js
    * Replace `fetchClustersAsArray` with `fetchClustersWithRetentionInfo` function call
    * Add `clusterIdToRetentionInfo`,  `hasReadWriteAccessForCluster` and `isDecommissionedClusterRetentionEnabled` properties to arg of `getColumnsForClusters` function call

8. Add src/Containers/Clusters/Components/ClusterDeletion.tsx

9. Edit src/Containers/Clusters/Components/ClusterSummary.tsx
    * Add `ClusterRetentionInfo` and `isDecommissionedClusterRetentionEnabled` props

10. Edit src/Containers/Clusters/cluster.helpers.tsx
    * Add `getClusterDeletionStatus` function

11. Edit src/Containers/Clusters/clusterTypes.ts
    * Re-export `Cluster` type from frontend cluster proto, as first step to supersede this container types file

12. Edit src/Containers/Clusters/clustersTableColumnDescriptors.js
    * Add `clusterIdToRetentionInfo`,  `hasReadWriteAccessForCluster` and `isDecommissionedClusterRetentionEnabled` properties
    * Update column width classes
    * Push **Cluster Deletion** column if feature flag is enabled
    * Push row actions if role has READ_WRITE_ACCESS for `'Cluster'` resource

13. Edit src/services/ClustersService.ts
    * Add `fetchClustersWithRetentionInfo` function
    * Rename `getClusterById` with `getClusterWithRetentionInforation` function

14. Edit src/types/cluster.proto.ts
    * Factor out `ClusterLabels`
    * Rename `ManagerType` as `ClusterManagerType`

15. Edit src/types/clusterService.proto.ts
    * Add `ClusterResponse`, `ClustersResponse`, `DecommissionedClusterRetentionInfo` types

### Residue

1. Replace imports from clusterTypes.ts with imports from frontend cluster proto
2. Rewrite files in TypeScript
3. Conditionally render other actions only if READ_WRITE_ACCESS for 'Cluster' resource

## Checklist

## Testing Performed

### manual testing

1. `export ROX_DECOMMISSIONED_CLUSTER_RETENTION=true` in ui also verified does not render if false
2. `yarn deploy-local` in ui
3. `yarn build` in ui
4. `yarn start` in ui

    * Visit /main/clusters: see GET /v1/clusters has new property

        ```js
        {
            clusterIdToRetentionInfo: {},
            clusters: […]
        }
        ```

    * Click cluster: see GET /v1/clusters/id has new property

        ```js
        {
            cluster: {…},
            clusterRetentionInfo: null
        }
        ```

### integration testing

1. `export CYPRESS_ROX_DECOMMISSIONED_CLUSTER_RETENTION=true` in ui/apps/platform
2. `yarn cypress-open` in ui/apps/platform
    * clusters.test.js
    * clusters/clusterDeletion.test.js see it skip tests without above feature flag export
